File: packages/quran_bridge_android/android/src/main/kotlin/dev/quran/bridge/QuranBridgePlugin.kt
package dev.quran.bridge

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

private const val METHOD_CHANNEL = "dev.quran.bridge/methods"
private const val AUDIO_CHANNEL  = "dev.quran.bridge/audio"

class QuranBridgePlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context

    private lateinit var service: QuranNativeService
    private lateinit var audioController: AudioController

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    // ── FlutterPlugin ────────────────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext

        service         = QuranNativeService(context)
        audioController = AudioController(context)

        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, AUDIO_CHANNEL)
        eventChannel.setStreamHandler(audioController)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        audioController.release()
        scope.cancel()
    }

    // ── MethodCallHandler ────────────────────────────────────────────────────

    override fun onMethodCall(call: MethodCall, result: Result) {
        scope.launch {
            try {
                val response = withContext(Dispatchers.IO) {
                    dispatch(call)
                }
                result.success(response)
            } catch (e: QuranBridgeException) {
                result.error(e.code, e.message, e.details)
            } catch (e: Exception) {
                result.error("UNKNOWN", e.message ?: "Unknown error", null)
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private suspend fun dispatch(call: MethodCall): Any? {
        val args = call.arguments as? Map<String, Any>

        return when (call.method) {
            "getSurahs"     -> service.getSurahs()

            "getPage"       -> {
                val page = args?.get("pageNumber") as? Int
                    ?: throw QuranBridgeException.invalidArgs("pageNumber")
                service.getPage(page)
            }

            "getVerses"     -> {
                val surah = args?.get("surah") as? Int
                    ?: throw QuranBridgeException.invalidArgs("surah")
                val from  = args["from"] as? Int
                val to    = args["to"]   as? Int
                service.getVerses(surah, from, to)
            }

            "addBookmark"   -> {
                val map = args ?: throw QuranBridgeException.invalidArgs("bookmark")
                service.bookmarkStore.add(map)
                null
            }

            "getBookmarks"  -> service.bookmarkStore.getAll()

            "removeBookmark" -> {
                val id = args?.get("id") as? String
                    ?: throw QuranBridgeException.invalidArgs("id")
                service.bookmarkStore.remove(id)
                null
            }

            "playAudio"     -> {
                val map = args ?: throw QuranBridgeException.invalidArgs("AudioRequest")
                withContext(Dispatchers.Main) { audioController.play(map) }
                null
            }

            "pauseAudio"    -> {
                withContext(Dispatchers.Main) { audioController.pause() }
                null
            }

            "stopAudio"     -> {
                withContext(Dispatchers.Main) { audioController.stop() }
                null
            }

            else -> throw UnsupportedOperationException("${call.method} not implemented")
        }
    }
}

File: packages/quran_bridge_ios/ios/Classes/QuranBridgePlugin.swift
import Flutter

private let kMethodChannel = "dev.quran.bridge/methods"
private let kAudioChannel  = "dev.quran.bridge/audio"

public class QuranBridgePlugin: NSObject, FlutterPlugin {

    private let service: QuranNativeService
    private let audioController: AudioController
    private var audioEventSink: FlutterEventSink?

    // MARK: - Registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: kMethodChannel,
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: kAudioChannel,
            binaryMessenger: registrar.messenger()
        )

        let instance = QuranBridgePlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    // MARK: - Init

    override init() {
        self.service = QuranNativeService()
        self.audioController = AudioController()
        super.init()
        self.audioController.delegate = self
    }

    // MARK: - MethodChannel dispatch

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]

        switch call.method {
        case "getSurahs":
            handleAsync(result: result) { try self.service.getSurahs() }

        case "getPage":
            guard let pageNumber = args?["pageNumber"] as? Int else {
                return result(invalidArgs("pageNumber"))
            }
            handleAsync(result: result) { try self.service.getPage(pageNumber) }

        case "getVerses":
            guard let surah = args?["surah"] as? Int else {
                return result(invalidArgs("surah"))
            }
            let from = args?["from"] as? Int
            let to   = args?["to"]   as? Int
            handleAsync(result: result) { try self.service.getVerses(surah: surah, from: from, to: to) }

        case "addBookmark":
            guard let map = args else { return result(invalidArgs("bookmark")) }
            handleAsync(result: result) { try self.service.bookmarkStore.add(from: map) }

        case "getBookmarks":
            handleAsync(result: result) { self.service.bookmarkStore.getAll() }

        case "removeBookmark":
            guard let id = args?["id"] as? String else {
                return result(invalidArgs("id"))
            }
            handleAsync(result: result) { self.service.bookmarkStore.remove(id: id) }

        case "playAudio":
            guard let map = args else { return result(invalidArgs("AudioRequest")) }
            handleAsync(result: result) { try self.audioController.play(from: map) }

        case "pauseAudio":
            handleAsync(result: result) { self.audioController.pause() }

        case "stopAudio":
            handleAsync(result: result) { self.audioController.stop() }

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Helpers

    /// Runs [body] on a background queue and marshals the result back to main.
    private func handleAsync(result: @escaping FlutterResult, body: @escaping () throws -> Any?) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let value = try body()
                DispatchQueue.main.async { result(value) }
            } catch let error as QuranBridgeError {
                DispatchQueue.main.async { result(error.flutterError) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "UNKNOWN", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func invalidArgs(_ field: String) -> FlutterError {
        FlutterError(code: "INVALID_ARGS", message: "Missing or invalid argument: \(field)", details: nil)
    }
}

// MARK: - EventChannel stream handler

extension QuranBridgePlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        audioEventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        audioEventSink = nil
        return nil
    }
}

// MARK: - AudioControllerDelegate

extension QuranBridgePlugin: AudioControllerDelegate {
    func audioController(_ controller: AudioController, didEmit state: [String: Any]) {
        DispatchQueue.main.async {
            self.audioEventSink?(state)
        }
    }
}
